enhancements/etcd/scaling-etcd-with-raft-learners.md

```diff
@@ -0,0 +1,343 @@
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in
+  [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Over time as clusters live longer and workloads grow the ability to scale the control-plane and replace failed nodes
+becomes a critical part of the admins maintenance overhead. Today the `cluster-etcd-operator` manages scaling up of
+the etcd cluster. To provide the foundation for initiatives such as scale down and vertical control plane scaling[1].
+The `cluster-etcd-operator` must ensure proper safety mechanisms exist to adjust membership of the etcd cluster.
+
+Introduced in etcd v3.4 the raft learner[2] provides mitigations which reduces quorum and
+stability issues during scaling. A learner is essentially an etcd member which is non voting thus can not impact 
+quorum but like other members will receive log replications from the leader.
+
+This enhancement proposes:
+- Replacing the default scale up performed by the cluster-etcd-operator to use raft learners.
+- Deprecation and removal of the current `discover-etcd-initial-cluster`[3] command and replacing its
+functionality with the existing `etcdenvvar` controller and subcommands of the `cluster-etcd-operator`. By
+eliminating this code it also removes the teams largest carry against upstream etcd.
+- Ensuring that scale up only happens on control-plane nodes which have a quorum-guard pod monitoring health.
+- Adding a flag to etcd which allows configuration of maximum learners in cluster (--max-learners). Today the max is 1.
+
+POC: Functional proof of concept: https://github.com/openshift/cluster-etcd-operator/pull/682
+
+[1] https://issues.redhat.com/browse/OCPPLAN-5712
+
+[2] https://etcd.io/docs/v3.4/learning/design-learner/
+
+[3] https://github.com/openshift/etcd/blob/openshift-4.10/openshift-tools/pkg/discover-etcd-initial-cluster/initial-cluster.go
+
+## Motivation
+
+Great demand exists to provide flexible automated scaling of the control-plane in the same way the worker
+nodes do today.
+
+### Goals
+
+- provide safe scale up of etcd cluster using raft learners.
+- add additional membership level details such as member ID to the init container validation process for etcd.
+- add observability to allow admin to clearly diagnose scaling failure.
+- reduce divergence from upstream etcd by removing `discover-etcd-initial-cluster`.
+
+### Non-Goals
+
+- implement scale down logic for the `cluster-etcd-operator`
+
+## Proposal
+
+### User Stories
+
+1. As a cluster-admin I want to scale up etcd without fear of quorum loss.
+
+2. As a cluster-admin I want to be able to replace failed control-place nodes without a deep understanding of etcd.
+
+### Critical Alerts
+
+We will add the following critical alerts:
+
+- alert about learner member which has not been promoted or started > 30s
```

Comment:
Is 30 seconds really the right threshold? We every "operator interval" initiate the learner promotion so should be at least "2 x operator interval" I would say. Also, I would not say critical more like a warning alert to help debug issues, we will already alert if the cluster has etcd errors and latency which should cover these issues anyways, no?

Reply:
Yeah MTTR of 7GB data file could be closer to 2m. So yes I would say this is an arbitrary number (30s) that needs additional thought maybe something longer ~5m. The general observation is that members should quickly be promoted and if not it points to a possible scaling issue.

Reply:
We might want 2 seperate observations

- learner not started (this should happen as fast as the installer pod can add a new etcd spec)
- learner not promoted (this really can be depending on the size of the state file and the speed of networking)

Reply:
I see we have the learner_promote_failures and learner_promote_successes, we can just do a rate error alert there, which can be critical, wdyt?

Reply:
Agreed. Rate error alert with % of errors is the standard for these kinds of things, we already have those kinds of alerts for other components.

Reply:
> islearner for > 5m

Does this not depend on the size of the etcd data?